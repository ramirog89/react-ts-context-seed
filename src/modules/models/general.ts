export type IAction<T> = { type: T | null, payload?: any }

export type Nullable<T> = { [P in keyof T]: T[P] | null } | any;