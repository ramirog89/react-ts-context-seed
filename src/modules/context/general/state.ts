import { GeneralModel } from '../../models';

export interface IState {
  toastList: GeneralModel.IToast[];
};

export const initialState: IState = {
  toastList: []
}
