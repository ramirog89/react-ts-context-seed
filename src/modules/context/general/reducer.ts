import { ActionType } from './actions';
import { IState } from './state';
import { GeneralModel } from '../../models';

export const reducer = (state: IState, action: GeneralModel.IAction<ActionType>) => {
  switch (action.type) {
    case ActionType.ADD_TOAST:
      return { ...state, toastList: [...state.toastList, action.payload.toast] };
    case ActionType.REMOVE_TOAST:
      return { ...state, toastList: state.toastList.filter((toast) => action.payload.id !== toast.id) };
    default:
      return state;
  }
};

