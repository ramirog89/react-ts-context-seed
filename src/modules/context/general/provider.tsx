import { useMemo, createContext, useReducer, useContext } from 'react';

import { ENV } from '../../../constants';
import { GeneralModel } from '../../models';
import { IProviderProps } from '../rootState';
import { IState, initialState } from './state';
import { ActionType } from './actions';
import { reducer } from './reducer';

export const GeneralContext = createContext<
  {
    state: IState;
    addToast: (todo: GeneralModel.IToast) => void;
  } | null
>(null);

export const useGeneral = () => {
  const context = useContext(GeneralContext);
  if (!context) {
    throw new Error('GeneralContext is not wrapped with GeneralContext.Provider');
  }
  return context;
};

export const GeneralProvider = (props: IProviderProps) => {
  const [state, dispatch] = useReducer(reducer, initialState);

  const addToast = (toast: GeneralModel.IToast) => {
    const id = new Date().toString();
    dispatch({ type: ActionType.ADD_TOAST , payload: { toast: { ...toast, id } } });
    setTimeout(() => {}, ENV.TOAST_DELAY);
    dispatch({ type: ActionType.REMOVE_TOAST, payload: { id } });
  };

  const value = useMemo(() => ({ state, addToast }), [state]);

  return <GeneralContext.Provider value={value} {...props} />
};

