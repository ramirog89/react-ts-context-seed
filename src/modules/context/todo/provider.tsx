import { useMemo, createContext, useReducer, useContext } from 'react';

import { GeneralModel, TodoModel } from '../../models';
import { IProviderProps } from '../rootState';
import { IState, initialState } from './state';
import { ActionType } from './actions';
import { reducer } from './reducer';
import { useGeneral } from '../general';

interface ITodoContext {
  state: IState;
  addTodo: (todo: TodoModel.ITodo) => void;
  removeTodo: (id: string) => void;
  fetchTodoList: () => void;
}

export const TodoContext = createContext<ITodoContext | null>(null);

export const useTodo = () => {
  const context = useContext(TodoContext);
  if (!context) {
    throw new Error('TodoContext is not wrapped with TodoContext.Provider');
  }
  return context;
};

export const TodoProvider = (props: IProviderProps) => {
  const { deps } = props;
  const [state, dispatch] = useReducer(reducer, initialState);
  const { addToast } = useGeneral();

  const fetchTodoList = async () => {
    try {
      const list = await deps.apiService.getTodo();
      dispatch({ type: ActionType.ADD_TODO_LIST, payload: { list } });
    } catch (e) {
      addToast({ message: 'error', type: GeneralModel.ToastType.ERROR });
    }
  }

  const addTodo = (todo: TodoModel.ITodo) => {
    const id = new Date().toString();
    dispatch({ type: ActionType.ADD_TODO , payload: { todo: { ...todo, id } } });
    addToast({ message: 'success', type: GeneralModel.ToastType.SUCCESS });
  };

  const removeTodo = (id: string) => {
    dispatch({ type: ActionType.REMOVE_TODO , payload: { id } });
  };

  const value = useMemo(() => ({ state, addTodo, removeTodo, fetchTodoList }), [state]); // eslint-disable-line

  return <TodoContext.Provider value={value} {...props} />
};

