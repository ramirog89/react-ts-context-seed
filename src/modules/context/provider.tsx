import React from 'react';
import { ApiService } from '../services/ApiService';

import { ComposeProvider, Provider } from './compose';
import { IDependencies } from './rootState';
import { TodoProvider } from './todo/provider';

export const dependencies: IDependencies = {
  apiService: new ApiService(),
};

export const providers: Provider[] = [TodoProvider];

export const ContextProvider = ({ children }: { children: React.ReactNode }) => {
  return (
    <ComposeProvider deps={dependencies} providers={providers}>
      {children}
    </ComposeProvider>
  )
}