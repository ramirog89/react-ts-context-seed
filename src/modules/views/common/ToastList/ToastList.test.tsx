import { getToast_1 } from '../../../../test/entities';
import { render } from '../../../../test/render';

import { GeneralContext } from '../../../context/general/provider';
import ToastList from './ToastList';

describe('ToastList', () => {
  let wrapper: any;

  it('should render', () => {
    wrapper = render(
      <GeneralContext.Provider value={{ state: { toastList: [getToast_1()] }, addToast: jest.fn() }}>
        <ToastList />
      </GeneralContext.Provider>
    );

    expect(wrapper.container).toMatchSnapshot();
  });

});
