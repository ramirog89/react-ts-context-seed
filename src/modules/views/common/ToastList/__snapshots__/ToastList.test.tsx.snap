// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToastList should render 1`] = `
<div>
  <div
    class="toast-list"
    style="position: fixed; bottom: 1rem; right: 1rem;"
  >
    <div
      class="success"
      style="border: 1px solid gray; padding: 6px; border-radius: 5px;"
    >
      todo test 1
    </div>
  </div>
</div>
`;
